src/librustc_mir/interpret/eval_context.rs

```diff
@@ -289,10 +290,8 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tc
         &self,
         frame: &Frame<'mir, 'tcx, M::PointerTag, M::FrameExtra>,
         local: mir::Local
-    ) -> EvalResult<'tcx, TyLayout<'tcx>> {
-        let local_ty = frame.mir.local_decls[local].ty;
-        let local_ty = self.monomorphize(local_ty, frame.instance.substs);
-        self.layout_of(local_ty)
+    ) -> TyLayout<'tcx> {
+        frame.local_layouts[local]
```

Comment:
While definitely an improvement for all currently possible const eval code, I think this is suboptimal for code with branching, as a lot of locals will never be touched during evaluation. I think it would be better to make `local_layouts` be `IndexVec<mir::Local, Option<TyLayout<'tcx>>>` and only fill it in on demand. To make this really work with reading I presume it would even need to be `IndexVec<mir::Local, RefCell<Option<TyLayout<'tcx>>>>` at which point we can just move to `IndexVec<mir::Local, Once<TyLayout<'tcx>>>` or some non-Sync variant of that.